pbj-core/pbj-runtime/src/main/java/com/hedera/pbj/runtime/io/buffer/RandomAccessData.java

```diff
@@ -567,4 +570,67 @@ default boolean contains(final long offset, @NonNull final RandomAccessData data
      * @param length The number of bytes to extract.
      */
     void writeTo(@NonNull final OutputStream outStream, final int offset, final int length);
+
+    /**
+     * Throws {@code IndexOutOfBoundsException} if the given {@code offset} is negative
+     * or greater than/equal to the given {@code length}.
+     *
+     * @param offset an offset in this RandomAccessData that the caller wants to access
+     * @param length the maximum offset plus one (1) that is allowed to be accessed.
+     *               It may be equal to the total length of the underlying buffer,
+     *               or be less than the total length when a limit value is used.
+     */
+    default void checkOffset(final long offset, final long length) {
+        if (offset < 0 || offset >= length) {
+            throw new IndexOutOfBoundsException("offset " + offset + " is out of bounds for length " + length);
+        }
+    }
+
+    /**
+     * Throws {@code IndexOutOfBoundsException} if the given {@code offset} is negative
+     * or greater than/equal to the given {@code length}, and throws {@code BufferUnderflowException}
+     * if the {@code offset + dataLength - 1} exceeds the given {@code length}.
+     * <p>
+     * checkOffsetToRead(offset, length, 1) is equivalent to checkOffset(offset, length)
+     * because the dataLength condition is always satisfied for dataLength == 1 byte
+     * as long as the initial offset is within the bounds.
+     *
+     * @param offset an offset in this RandomAccessData that the caller wants to access
+     * @param length the maximum offset plus one (1) that is allowed to be accessed.
+     *               It may be equal to the total length of the underlying buffer,
+     *               or be less than the total length when a limit value is used.
+     * @param dataLength the length of the data to read
+     */
+    default void checkOffsetToRead(final long offset, final long length, final long dataLength) {
+        if (offset < 0 || offset > length || (offset == length && dataLength != 0)) {
+            throw new IndexOutOfBoundsException("offset " + offset + " is out of bounds for length " + length);
+        }
+        if (offset + dataLength - 1 >= length) {
```

Comment:
`offset + dataLength - 1 >= length`
==
`offset + dataLength > length`
==
`offset > length - dataLength`

The latter form is preferable to avoid overflows.

Reply:
Thanks for the tip. I've updated it.